File: src/main/java/Field.java
import java.util.ArrayList;

public class Field {
    public boolean[][] grid;
    public int width;
    public int height;

    public Field(int width, int height) {
        this.width = width;
        this.height = height;
        this.grid = new boolean[height][width];
    }

    public boolean[][] getGrid() { return this.grid; }

    private boolean isAvailable(int line, int column) {
        if (line < 0 || line >= this.grid.length) return false;
        if (column < 0 || column >= this.grid[0].length) return false;
        return !this.grid[line][column];
    }
    
    public boolean isAvailable(Tromino t, int line, int column, int rotation) {
        int[][] offsets = t.getOffsets(rotation);
        for (int[] off : offsets) {
            int l = line + off[0];
            int c = column + off[1];
            if (!isAvailable(l, c)) return false;
        }
        return true;
    }
    /**
     * Place a tromino at (line, column) using the given rotation.
     * Rotation semantics are provided by each Tromino subclass via getOffsets(rotation).
     * Returns true if placement succeeded, false on collision or out-of-bounds.
     */
    public boolean addTromino(Tromino t, int line, int column, int rotation) {
        int[][] offsets = t.getOffsets(rotation);
        // validate all offsets first
        if (!isAvailable(t, line, column, rotation)) {
            return false;
        }
        // commit placement
        for (int[] off : offsets) {
            int l = line + off[0];
            int c = column + off[1];
            this.grid[l][c] = true;
        }
        return true;
    }

    public ArrayList<Point> getEmptyPositions() {
        ArrayList<Point> positions = new ArrayList<>();
        for (int l = 0; l < this.grid.length; l++) {
            for (int c = 0; c < this.grid[0].length; c++) {
                if (!this.grid[l][c]) {
                    positions.add(new Point(l, c));
                }
            }
        }
        return positions;
    }

    @Override
    public Field clone() {
        Field newField = new Field(this.grid[0].length, this.grid.length);
        for (int l = 0; l < this.grid.length; l++) {
            for (int c = 0; c < this.grid[0].length; c++) {
                newField.grid[l][c] = this.grid[l][c];
            }
        }
        return newField;
    }
}

File: src/Field.cpp
#include "Field.h"

bool Field::isAvailable(int line, int column) const
{
    if (line < 0 || line >= height_)
        return false;
    if (column < 0 || column >= width_)
        return false;
    return !grid_[line][column];
}

bool Field::isAvailable(const Tromino& t, int line, int column, int rotation) const
{
    std::vector<Offset> offsets = t.getOffsets(rotation);
    for (const Offset& off : offsets)
    {
        int l = line + off[0];
        int c = column + off[1];
        if (!isAvailable(l, c))
            return false;
    }
    return true;
}

bool Field::addTromino(const Tromino& t, int line, int column, int rotation)
{
    if (!isAvailable(t, line, column, rotation))
        return false;
    std::vector<Offset> offsets = t.getOffsets(rotation);
    for (const Offset& off : offsets)
    {
        int l = line + off[0];
        int c = column + off[1];
        grid_[l][c] = true;
    }
    return true;
}

std::vector<Point> Field::getEmptyPositions() const
{
    std::vector<Point> positions;
    for (int l = 0; l < height_; ++l)
    {
        for (int c = 0; c < width_; ++c)
        {
            if (!grid_[l][c])
                positions.emplace_back(l, c);
        }
    }
    return positions;
}

Field Field::clone() const
{
    Field clone(grid_);
    return clone;
}

std::ostream& operator<<(std::ostream& os, const Field& f)
{

    for (int l = 0; l < f.getHeight(); ++l)
    {
        for (int c = 0; c < f.getWidth(); ++c)
        {
            os << (f.getGrid()[l][c] ? '*' : '.');
        }
        os << '\n';
    }
    return os;
}

std::vector<int> Field::getColumnHeights() const
{
    std::vector<int> heights(width_, 0);
    for (int c = 0; c < width_; ++c)
    {
        for (int r = 0; r < height_; ++r)
        {
            if (grid_[r][c])
            {
                heights[c] = height_ - r;
                break;
            }
        }
    }
    return heights;
}

int Field::getAggregateHeight() const
{
    std::vector<int> heights = getColumnHeights();
    int totalHeight = 0;
    for (int h : heights)
    {
        totalHeight += h;
    }
    return totalHeight;
}

int Field::getCompleteLines() const
{
    int lines = 0;
    for (int r = 0; r < height_; ++r)
    {
        bool isComplete = true;
        for (int c = 0; c < width_; ++c)
        {
            if (!grid_[r][c])
            {
                isComplete = false;
                break;
            }
        }
        if (isComplete)
        {
            lines++;
        }
    }
    return lines;
}

int Field::getHoles() const
{
    int holes = 0;
    for (int c = 0; c < width_; ++c)
    {
        bool roof = false;
        for (int r = 0; r < height_; ++r)
        {
            if (grid_[r][c])
            {
                roof = true;
            }
            else if (roof)
            {
                holes++;
            }
        }
    }
    return holes;
}

int Field::getBumpiness() const
{
    std::vector<int> heights = getColumnHeights();
    int bumpiness = 0;
    for (int c = 0; c < width_ - 1; ++c)
    {
        bumpiness += std::abs(heights[c] - heights[c + 1]);
    }
    return bumpiness;
}
